[reland][DTensor][3/N] add DTensor constructor function: full (#101436)

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103165

This is a reland attempt of reverted PR #101436 .

Differential Revision: [D46537531](https://our.internmc.facebook.com/intern/diff/D46537531)

XilunWu committed:
[reland][DTensor][3/N] add DTensor constructor function: full (#101436)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/__init__.py:
```diff
@@ -157,6 +160,53 @@ def empty(
+        requires_grad (bool, optional): If autograd should record operations on the
+            returned :class:`DTensor`. Default: ``False``.
+        device_mesh: :class:`DeviceMesh` type, contains the mesh info of ranks.
+        placements: a sequence of :class:`Placement` type: ``Shard``, ``Replicate``, ``_Partial``.
```
Actually we shouldn't have `_Partial` exposed to the doc, it's a private placement and we shouldn't document it in `full` API

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46537531).

XilunWu committed:
Update on "[reland][DTensor][3/N] add DTensor constructor function: full (#101436)"


This is a reland attempt of reverted PR #101436 .

Differential Revision: [D46537531](https://our.internmc.facebook.com/intern/diff/D46537531)

[ghstack-poisoned]

XilunWu commented:
@pytorchmergebot merge
